[E2E] Adds test harness (#589)
pokt-network/pocket: e2e/tests/README.md

> e2e tests
> =========
> 
> > tl; dr - `make localnet_up` and then `make test_e2e` 

Reviewer: ```suggestion
# E2E Tests

> tl; dr - `make localnet_up` and then `make test_e2e` 
```

> 

Reviewer: Consider using [1] to create a table of contents.

[1] https://marketplace.visualstudio.com/items?itemName=huntertran.auto-markdown-toc

> You can also click the `e2e-tests` button in the Tilt UI, which is handy during development.

Reviewer: Insert a screenshot?

## Design & Architecture

The test suite is located in `e2e/tests` and it contains a set of Cucumber Gherkin feature files and the associated Go tests to run them. This is a test file that lays out the BDD pattern using the client's help function as an example.

### Implementation 

> `make test_e2e` sees any files named with the pattern `*.feature` in `e2e/tests` and runs them with [godog](https://github.com/cucumber/godog), the Go test runner for Cucumber tests. We indirectly depend on the LocalNet to be up and running for our test suite to use. We call commands to steer around a container by calling `kubectl exec` and targeting a pod in the cluster by name. We are able to pass commands to that client and then assert on the behavior of that program.

Reviewer: ```suggestion
`make test_e2e` sees any files named with the pattern `*.feature` in `e2e/tests` and runs them with [godog](https://github.com/cucumber/godog), the Go test runner for Cucumber tests. There is an indirect dependance on the LocalNet to be up and running for our test suite to use. We call commands to steer around a container by calling `kubectl exec` and targeting a pod in the cluster by name. We are able to pass commands to that client and then assert on the behavior of that program.
```

Try to avoid using `we` and make it more 3rd person.

> ```
> e2e/
> └── tests
>     ├── README.md
>     ├── kube_client.go
>     ├── root.feature
>     └── steps_init_test.go
> 
> 2 directories, 4 files
> ```
> 
> `kube_client.go` defines the KubeClient that grabs a Pod and drives it around for the tests.
> `steps_init_test.go` registers the handler functions and runs the test suite.

Reviewer: This will highlight things appropriately.
```suggestion
```bash
e2e/
└── tests
    ├── README.md
    ├── kube_client.go # defines the KubeClient that grabs a Pod and drives it around for the tests.
    ├── root.feature # registers the handler functions and runs the test suite.
    └── steps_init_test.go

2 directories, 4 files
```

### KubeClient and Kubernetes Clientsets

> The KubeClient wraps the command and saves a reference to the last run command's result. The local environment's Kube client configuration is used to call commands. The Kubernetes Clientset can be changed in the test suite by swapping out the client that is defined at the package level. Multiple configurations can be maintained to test multiple environments and they can be swapped out arbitrarily.

Reviewer: @okdas Can I get your feedback here. I'm very weary of enforcing my own style, but I feel like this could be easier to read as 4 separate bullet points instead of a paragraph.

Looking for an extra opinion (outside my own and Dylan's).

### Separation of Concerns

> The hardest architectural problem is correctly waiting for services so that we can call the commands at the correct moment, but doing so in the package without altering the LocalNet component. We should indirectly depend on the LocalNet but not modify it, since we consider it a separate concern. To achieve this, we will need to query the Kubernetes Clientset that we acquire from the local configuration and repeatedly poll that for changes until all services our tests rely on are running.

Reviewer: this is good to know, but I feel like there is missing context.

I'm trying to put myself in the shoes of someone other than me or you (who has not been thinking about this problem). Is this TMI? What were your thoughts of the v0 E2E doc?
